pkg/services/tree/sync.go

```diff
@@ -0,0 +1,98 @@
+package tree
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/nspcc-dev/neofs-node/pkg/local_object_storage/pilorama"
+	"github.com/nspcc-dev/neofs-node/pkg/network"
+	cid "github.com/nspcc-dev/neofs-sdk-go/container/id"
+	"google.golang.org/grpc"
+)
+
+// Synchronize tries to synchronize log starting from the last stored height.
+func (s *Service) Synchronize(ctx context.Context, cid *cid.ID, treeID string) error {
+	nodes, err := s.getContainerNodes(cid)
+	if err != nil {
+		return err
+	}
+
+	lm, err := s.forest.TreeGetOpLog(cid, treeID, 0)
+	if err != nil && !errors.Is(err, pilorama.ErrTreeNotFound) {
+		return err
+	}
+
+	height := lm.Time + 1
+	for _, n := range nodes.Flatten() {
+		n.IterateAddresses(func(addr string) bool {
+			var a network.Address
+			if err := a.FromString(addr); err != nil {
+				return false
+			}
+
+			cc, err := grpc.DialContext(ctx, a.HostAddr(), grpc.WithInsecure())
+			if err != nil {
+				// Failed to connect, try the next address.
+				return false
```

Comment:
If I correctly understand we don't return an error if all addresses are "problematic", is it expected?

Reply:
Yes, we query every node from the container, some of them can be unavailable, do our best here.